feat: support other audio format

AnneYang720 opened the pull request:
Goals:

This PR is related to issue #998 
- [x] refactor `load()` method in AudioUrl to use PyDub
- [x] remove useless code
- [x] pass current tests
- [x] add tests for other format

JohannesMessner commented on docarray/typing/url/audio_url.py:
```diff
@@ -66,9 +66,16 @@ class MyDoc(Document):
+        samples = segment.get_array_of_samples()
+        samples = np.array(samples)
+
+        samples_norm = samples / 2 ** (segment.sample_width * 8 - 1)
```
I have no idea about audio data so maybe this is a stupid question, but what are these numbers here? Are they guaranteed to be always these same numbers, no matter the data? And what does this transformation do?

AnneYang720 replied:
Line 74 and 75 are how to get ndarray from audio file with pydub AudioSegment.

JohannesMessner replied:
okk I see. Maybe we should have a comment like `# normalize amplitude to [-1, 1]`?

samsja commented on .github/workflows/ci.yml:
```diff
@@ -210,6 +214,7 @@ jobs:
           poetry install --all-extras
           poetry run pip install protobuf==3.19.0
           poetry run pip install tensorflow==2.11.0
+          sudo apt-get install --no-install-recommends ffmpeg
```
`--no-install-recommends` what is this doing ?

AnneYang720 replied:
Don't let apt to install some "recommended" (not needed) packages, to save a little bit of CI time.

samsja commented on docarray/typing/url/audio_url.py:
```diff
@@ -66,9 +44,14 @@ class MyDoc(Document):
+        samples = np.array(segment.get_array_of_samples())
 
-        bytes_ = AudioBytes(self.load_bytes())
-        return bytes_.load()
```
good point

AnneYang720 replied:
The problem is `AudioBytes` and its `load()` method are not currently used anywhere.

AnneYang720 commented:
I added the function `save_to_audio_file`. During the development, I found the `frame_rate` of an audio file is essential, so I added an variable `frame_rate` for `class Audio`. So now it is
```python
audio = Audio(
    url='https://github.com/docarray/docarray/blob/feat-rewrite-v2/tests/toydata/hello.wav?raw=true'
)
audio.tensor, audio.frame_rate = audio.url.load()
```
just like in video we have
```python
vid.audio.tensor, vid.tensor, vid.key_frame_indices = vid.url.load()
```

It may not be very compact. Maybe I can wrap the return to a class like `VideoLoadResult`.

samsja commented on docarray/typing/tensor/audio/abstract_audio_tensor.py:
```diff
@@ -43,6 +43,28 @@ def save_to_wav_file(
             f.setframerate(sample_rate)
             f.writeframes(self.to_bytes())
 
+    def save_to_audio_file(
```
yes

samsja reviewed (changes requested):
add sample rate to audio predefined doc

samsja commented:
> 
    * simply have a huge-ass list of all supported file extensions, move it to a different file if needed
    * if a path has an extension and it is not in that list, then fail
    * if a path does not have an extension, pass
  
I like this tradeoff. 

The goal of validation is mostly to catch if someone put an `file.png` instead of a `file.wav`.

samsja commented on docarray/typing/tensor/audio/abstract_audio_tensor.py:
```diff
@@ -20,28 +19,36 @@ def to_bytes(self):
+        format: str = 'wav',
+        frame_rate: int = 44100,
         sample_width: int = 2,
+        *args,
```
I know we had this pattern a lot in v1, but I am not sure i lile it tbh.

What about having a `pydub_args: Dict[str, Any]` parameters

and do segment.export(..., **pydub_arg)

samsja replied:
@JohannesMessner what do u think ?

samsja reviewed (changes requested):
looks great, one comment about the save function

samsja commented on docarray/typing/tensor/audio/abstract_audio_tensor.py:
```diff
@@ -20,28 +19,35 @@ def to_bytes(self):
+        format: str = 'wav',
+        frame_rate: int = 44100,
         sample_width: int = 2,
+        pydub_args: Dict[str, Any] = {},
```
can u add the corresponding docstring as well ? Just say that this arg are forwarded to the export function
